client.go

```diff
@@ -21,74 +21,168 @@ import (
 	}
-	if err := httpsClient.initClientCert(args); err != nil {
+	return client, nil
+}
+
+func NewHttpClient() *httpClient {
+	var client httpClient
+	client.minImageSize = minimumImageSize
+	client.HTTPClient = &http.Client{}
+
+	return &client
+}
+
+func NewHttpsClient(conf httpsClientConfig) *httpsClient {
+	var client httpsClient
+	client.httpClient = *NewHttpClient()
+
+	if err := client.initServerTrust(conf); err != nil {
+		log.Error("Can not initialize server trust: ", err)
 		return nil
 	}
 
-	tlsConf := tls.Config{
-		RootCAs:      &httpsClient.trustedCerts,
-		Certificates: []tls.Certificate{httpsClient.clientCert},
+	if err := client.initClientCert(conf); err != nil {
+		log.Error("Can not initialize client certificate: ", err)
+		return nil
 	}
 
 	transport := http.Transport{
-		TLSClientConfig: &tlsConf,
+		TLSClientConfig: &tls.Config{
+			RootCAs:      &client.trustedCerts,
+			Certificates: []tls.Certificate{client.clientCert},
+		},
 	}
 
-	httpsClient.HTTPClient = &http.Client{
-		Transport: &transport,
+	client.HTTPClient.Transport = &transport
+	return &client
+}
+
+// Client configuration
+
+type httpsClientConfig struct {
+	certFile   string
+	certKey    string
+	serverCert string
+	isHttps    bool
+}
+
+type httpsClientAuthCreds struct {
+	// Cert+privkey that authenticates this client
+	clientCert tls.Certificate
+	// Trusted server certificates
+	trustedCerts x509.CertPool
+}
+
+func (c *httpsClient) initServerTrust(conf httpsClientConfig) error {
+	if conf.serverCert == "" {
+		// TODO: this is for pre-production version only to simplify tests.
+		// Make sure to remove in production version.
+		log.Warn("Server certificate not provided. Trusting all servers.")
+		return nil
+	}
+
+	c.trustedCerts = *x509.NewCertPool()
+	// Read certificate file.
+	cacert, err := ioutil.ReadFile(conf.serverCert)
+	if err != nil {
+		return err
 	}
+	c.trustedCerts.AppendCertsFromPEM(cacert)
 
-	return &httpsClient
+	if len(c.trustedCerts.Subjects()) == 0 {
+		return errorAddingServerCertificateToPool
+	}
+	return nil
 }
 
-func (c *client) GetScheduledUpdate(process RequestProcessingFunc, server string) (interface{}, error) {
-	r, err := c.makeAndSendRequest(http.MethodGet, server)
+func (c *httpsClient) initClientCert(conf httpsClientConfig) error {
+	if conf.certFile == "" || conf.certKey == "" {
+		// TODO: this is for pre-production version only to simplify tests.
+		// Make sure to remove in production version.
+		log.Warn("No client key and certificate provided. Using auto-generated.")
+		return nil
+	}
+
+	clientCert, err := tls.LoadX509KeyPair(conf.certFile, conf.certKey)
 	if err != nil {
+		return errorLoadingClientCertificate
+	}
+	c.clientCert = clientCert
+	return nil
+}
+
+func (c *httpClient) GetScheduledUpdate(process RequestProcessingFunc,
+	server string, deviceID string) (interface{}, error) {
+	// http client should be able to perform https requests
+	if strings.HasPrefix(server, "http") {
```

Comment:
That's a good point. Actually this is still bit tricky as we are not having the server and we are using 3rd party http or https services. I am sure this will be changed in future, but anyway I will change to check for either ```http://``` or ```https://```.